src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/SocksHelper.cs

```diff
@@ -0,0 +1,339 @@
+                // |VER | CMD |  RSV  | ATYP | DST.ADDR | DST.PORT |
+                // +----+-----+-------+------+----------+----------+
+                // | 1  |  1  | X'00' |  1   | Variable |    2     |
+                // +----+-----+-------+------+----------+----------+
+                buffer[0] = ProtocolVersion5;
+                buffer[1] = CMD_CONNECT;
+                buffer[2] = 0;
+                buffer[3] = ATYP_DOMAIN_NAME;
+
+                int addressLength = Encoding.UTF8.GetByteCount(host);
+                buffer[4] = checked((byte)addressLength);
+                int bytesEncoded = Encoding.UTF8.GetBytes(host, buffer.AsSpan(5));
+                Debug.Assert(bytesEncoded == addressLength);
+
+                Debug.Assert(port > 0);
+                Debug.Assert(port < ushort.MaxValue);
+                buffer[addressLength + 5] = (byte)(port >> 8);
+                buffer[addressLength + 6] = (byte)port;
+
+                if (async)
+                {
+                    await stream.WriteAsync(buffer.AsMemory(0, addressLength + 7)).ConfigureAwait(false);
+                }
+                else
+                {
+                    stream.Write(buffer.AsSpan(0, addressLength + 7));
+                }
+
+                // +----+-----+-------+------+----------+----------+
+                // |VER | REP |  RSV  | ATYP | DST.ADDR | DST.PORT |
+                // +----+-----+-------+------+----------+----------+
+                // | 1  |  1  | X'00' |  1   | Variable |    2     |
+                // +----+-----+-------+------+----------+----------+
+                await ReadToFillAsync(stream, buffer.AsMemory(0, 4), async).ConfigureAwait(false);
+                if (buffer[0] != ProtocolVersion5)
+                    throw new Exception("Bad protocol version");
+                if (buffer[1] != REP_SUCCESS)
+                    throw new Exception("Connection failed");
+                switch (buffer[3])
+                {
+                    case ATYP_IPV4:
+                        await ReadToFillAsync(stream, buffer.AsMemory(0, 4), async).ConfigureAwait(false);
+                        break;
+                    case ATYP_IPV6:
+                        await ReadToFillAsync(stream, buffer.AsMemory(0, 16), async).ConfigureAwait(false);
+                        break;
+                    case ATYP_DOMAIN_NAME:
+                        await ReadToFillAsync(stream, buffer.AsMemory(0, 1), async).ConfigureAwait(false);
+                        addressLength = buffer[0];
+                        await stream.ReadAsync(buffer.AsMemory(0, addressLength)).ConfigureAwait(false);
+                        break;
+                    default:
+                        throw new Exception("Unknown address type");
+                }
+                await ReadToFillAsync(stream, buffer.AsMemory(0, 2), async).ConfigureAwait(false);
+                // response address not used
```

Comment:
This part is essentially calculating how many bytes to skip. It could written as:

```c#
// reading 7 bytes instead of 4 - includes the first byte of address and the port
await ReadToFillAsync(stream, buffer.AsMemory(0, 7), async).ConfigureAwait(false);

if (buffer[0] != ProtocolVersion5)
// ...

int bytesToSkip = buffer[3] switch
{
    ATYP_IPV4 => 4 - 1,
    ATYP_IPV6 => 16 - 1,
    ATYP_DOMAIN_NAME => buffer[4],
    _ => throw new Exception("Unknown address type received")
};

await ReadToFillAsync(stream, buffer.AsMemory(0, bytesToSkip), async).ConfigureAwait(false);
```